hoomd/md/pytest/test_rigid.py

```diff
@@ -0,0 +1,242 @@
+        current_body_definition[key] = valid_body_definition[key]
+
+
+def check_bodies(snapshot, definition):
+    """Non-general assumes a snapshot from two_particle_snapshot_factory.
+
+    This is just to prevent duplication of code from test_create_bodies and
+    test_running_simulation.
+    """
+    assert snapshot.particles.N == 10
+    assert all(snapshot.particles.typeid[3:] == 1)
+
+    assert snapshot.particles.body[0] == 0
+    assert all(snapshot.particles.body[2:6] == 0)
+    assert snapshot.particles.body[1] == 1
+    assert all(snapshot.particles.body[6:] == 1)
+
+    # check charges
+    for i in range(4):
+        assert snapshot.particles.charge[i + 2] == definition["charges"][i]
+        assert snapshot.particles.charge[i + 6] == definition["charges"][i]
+
+    # check diameters
+    for i in range(4):
+        assert snapshot.particles.diameter[i + 2] == definition["diameters"][i]
+        assert snapshot.particles.diameter[i + 6] == definition["diameters"][i]
+
+    particle_one = (
+        snapshot.particles.position[0], snapshot.particles.orientation[0])
+    particle_two = (
+        snapshot.particles.position[1], snapshot.particles.orientation[1])
+
+    # Check positions
+    def check_position(central_position, central_orientation,
+                       constituent_position, local_position):
+        d_pos = rowan.rotate(central_orientation, local_position)
+        assert np.allclose(central_position + d_pos, constituent_position)
+
+    for i in range(4):
+        check_position(*particle_one, snapshot.particles.position[i + 2],
+                       definition["positions"][i])
+        check_position(*particle_two, snapshot.particles.position[i + 6],
+                       definition["positions"][i])
+
+    # check orientation
+    def check_orientation(central_orientation, constituent_orientation,
+                          local_orientation):
+        expected_orientation = rowan.normalize(
+            rowan.multiply(central_orientation, local_orientation))
+        assert np.allclose(expected_orientation, local_orientation)
+
+    for i in range(4):
+        check_orientation(particle_one[1],
+                          snapshot.particles.orientation[i + 2],
+                          definition["orientations"][i])
+        check_orientation(particle_two[1],
+                          snapshot.particles.orientation[i + 6],
+                          definition["orientations"][i])
+
+
+def test_create_bodies(simulation_factory, two_particle_snapshot_factory,
+                       valid_body_definition):
+    rigid = md.constrain.Rigid()
+    rigid.body["A"] = valid_body_definition
+
+    initial_snapshot = two_particle_snapshot_factory()
+    initial_snapshot.particles.types = ["A", "B"]
+    sim = simulation_factory(initial_snapshot)
+
+    rigid.create_bodies(sim.state)
+    snapshot = sim.state.snapshot
+    if sim.device.communicator.rank == 0:
```

Comment:
Did we get rid of the `snaphost.exists` property? I believe this line is the case where we would normally use that.